### PR TITLE
9179 add default source to all layers

### LINKED ISSUE #9179: If you create a map (with several layers), and then hide any of them, map fails
## STR

- Create a map with several layers (try [this](http://withoutbaggage.com/gps/cycling-vietnam-without-baggage.gpx)).
- Once in the map, with 3 layers, try to hide one of them.
- Map fails due to source X0 doesn't exist.

In my opinion, this is another clue that we need to create and store the source analysis from the beginning in the backend when a new map is created, with that we would avoid all the hacks we are doing in the frontend side and more bugs that will appear.

PS: styling also fails at the first time you try.

@CartoDB/backend @CartoDB/frontend

### PR DESCRIPTION
**Pull request by gfiorav**
Fixes #9179 

This achieves that Builder maps (Builder maps only) are ensured to have a source analysis per layer if the don't have any other type analysis already.

This should be wiped from the face of the code once the analysis logic has been ported to the backend.

Please CR, @javitonino

### CODE REVIEW

File: app/models/carto/analysis.rb [hunk 1]
```diff
@@ -29,6 +29,24 @@ def self.find_by_natural_id(visualization_id, natural_id)
     analysis
   end
 
+  def self.source_analysis_for_layer(layer, index)
+    map = layer.map
+    visualization = map.visualization if map
+    user = visualization.user if visualization
+
+    visualization_id = visualization.id if visualization
+    user_id = user.id if user
```
Comment (javitonino) [inline app/models/carto/analysis.rb]: The code above is maybe too defensive and not too useful, in the sense that if one of the conditionals fail, `user_id` will be set to nil and the creation at the bottom of this method will fail anyway.

File: app/models/carto/analysis.rb [hunk 2]
```diff
@@ -29,6 +29,24 @@ def self.find_by_natural_id(visualization_id, natural_id)
...
+
+    analysis_definition = {
+      id: 'abcdefghijklmnopqrstuvwxyz'[index] + '0',
+      type: 'source',
+      params: { query: layer.default_query(user) },
+      options: { table_name: layer.options[:table_name] }
+    }
+
+    new(visualization_id: visualization_id, user_id: user_id, analysis_definition: analysis_definition)
```
Comment (javitonino) [inline app/models/carto/analysis.rb]: How come this saves the analysis without calling `save` or `create`?

**Comment by javitonino** [comment]
I left some comments, one of them may need some attention. Also, I would love to see some tests for this (at least the common case for a migration with a couple of layers).

**Comment by javitonino** [comment]
Oh, and about the SQL thing, I think the frontend detects this by looking at the query and matching it to "SELECT * FROM `table_name`". It may fail due to capitalization or adding the username. I'm dropping a comment to Nacho on his PR (risking my life with this) as I think it was an issue on that one as well.

File: app/models/carto/analysis.rb (hunk 2 above)
Comment (gfiorav) [inline app/models/carto/analysis.rb]: It doesn't! It's a builder, it returns the model _built_ and then you save it or whatever.

Up for debate, of course.

File: app/models/carto/analysis.rb (hunk 1 above)
Comment (gfiorav) [inline app/models/carto/analysis.rb]: Yeah, I agree.

File: app/models/carto/analysis.rb (hunk 2 above)
Comment (javitonino) [inline app/models/carto/analysis.rb]: Then I don't see where in this PR it is being saved :confused:

**Comment by gfiorav** [comment]
Last CR please!

**Comment by javitonino** [comment]
:+1: Be careful with that query change though, but it looks like the right thing to do.

**Pull request merged**